client/state/routing/middleware.js

```diff
@@ -0,0 +1,37 @@
+/**
+ * External dependencies
+ */
+import debugFactory from 'debug';
+import page from 'page';
+import store from 'store';
+
+/**
+ * Internal dependencies
+ */
+import { ROUTE_SET } from 'state/action-types';
+
+const debug = debugFactory( 'calypso:restore-last-location' );
+
+let hasInitialized = false;
+
+export const restoreLastLocation = () => ( next ) => ( action ) => {
```

Comment:
@mcsf Isn't this signature normally `export const restoreLastLocation = store => next => action => {`